lib/ast/AST.cpp

```diff
@@ -928,6 +928,110 @@ sptr<KOREPattern> KORECompositePattern::filterSubstitution(PrettyPrintData const
+      i++;
+      continue;
+    }
+    auto axiom = dynamic_cast<KOREAxiomDeclaration *>(decl.get());
+    auto equals = dynamic_cast<KORECompositePattern *>(axiom->getPattern().get());
+    auto lhs = equals->arguments[reverse ? 1 : 0];
+    auto rhs = equals->arguments[reverse ? 0 : 1];
+    substitution subst;
+    bool matches = lhs->matches(subst, subsorts, overloads, shared_from_this());
+    if (matches && (decl->getAttributes().count("macro-rec") || decl->getAttributes().count("alias-rec") || !appliedRules.count(i))) {
+      std::set<size_t> oldAppliedRules = appliedRules;
+      appliedRules.insert(i);
+      auto result = rhs->substitute(subst)->expandMacros(subsorts, overloads, macros, reverse, appliedRules);
+      appliedRules = oldAppliedRules;
+      return result;
+    }
+    i++;
+  }
+  if (arguments.empty()) {
+    return shared_from_this();
+  }
+  sptr<KORECompositePattern> result = KORECompositePattern::Create(constructor.get());
+  for (auto &arg : arguments) {
+    result->addArgument(arg->expandMacros(subsorts, overloads, macros, reverse));
+  }
+  return result;
+}
+
+bool KORECompositePattern::matches(substitution &subst, SubsortMap const& subsorts, SymbolMap const& overloads, sptr<KOREPattern> subject) {
+  auto subj = dynamic_cast<KORECompositePattern *>(subject.get());
+  if (!subj) {
+    return false;
+  }
+  if (*subj->getConstructor() != *getConstructor()) {
+    if (subj->getConstructor()->getName() == "inj" && getConstructor()->getName() == "inj") {
+      if (*subj->getConstructor()->getFormalArguments()[1] != *getConstructor()->getFormalArguments()[1]) {
+        return false;
+      }
+      sptr<KORESort> a = subj->getConstructor()->getFormalArguments()[0];
+      sptr<KORESort> b = getConstructor()->getFormalArguments()[0];
+      if (subsorts.count(b.get()) && subsorts.at(b.get()).count(a.get())) {
+        sptr<KORECompositePattern> ba = KORECompositePattern::Create("inj");
+        ba->getConstructor()->addArgument(b);
+        ba->getConstructor()->addArgument(a);
+        ba->addArgument(arguments[0]);
+        return ba->matches(subst, subsorts, overloads, subj->getArguments()[0]);
+      } else if (subsorts.count(a.get()) && subsorts.at(a.get()).count(b.get())) {
+        sptr<KORECompositePattern> ab = KORECompositePattern::Create("inj");
+        ab->getConstructor()->addArgument(a);
+        ab->getConstructor()->addArgument(b);
+        ab->addArgument(subj->getArguments()[0]);
+        return arguments[0]->matches(subst, subsorts, overloads, ab);
+      } else {
+        return false;
+      }
+    } else if (subj->getConstructor()->getName() == "inj") { 
+      sptr<KOREPattern> child = subj->getArguments()[0];
+      if (auto composite = dynamic_cast<KORECompositePattern *>(child.get())) {
+        if (overloads.count(composite->getConstructor()) && overloads.at(composite->getConstructor()).count(getConstructor())) {
```

Comment:
Does it make sense to also check for this overloads stuff in `this`, i.e. not only in `subj`?

Reply:
This is an accurate representation of how matching works in the llvm backend. I don't think it's appropriate to explore changed behavior here.